aggregator/aggregator.go

```diff
@@ -120,26 +129,57 @@ func NewMocked(fun, regex, prefix, sub, outFmt string, cache bool, interval, wai
 	if cache {
 		a.reCache = make(map[string]CacheEntry)
 	}
+	a.setKey()
+	a.numIn = stats.Counter("unit=Metric.direction=in.aggregator=" + a.Key)
+	a.numFlushed = stats.Counter("unit=Metric.direction=out.aggregator=" + a.Key)
 	a.wg.Add(1)
 	go a.run()
 	return a, nil
 }
 
-type aggkey struct {
-	key string
-	ts  uint
+type TsSlice []uint
+
+func (p TsSlice) Len() int           { return len(p) }
+func (p TsSlice) Less(i, j int) bool { return p[i] < p[j] }
+func (p TsSlice) Swap(i, j int)      { p[i], p[j] = p[j], p[i] }
+
+func (a *Aggregator) setKey() string {
+	h := md5.New()
+	h.Write([]byte(a.Fun))
+	h.Write([]byte("\000"))
+	h.Write([]byte(a.Regex))
+	h.Write([]byte("\000"))
+	h.Write([]byte(a.Prefix))
+	h.Write([]byte("\000"))
+	h.Write([]byte(a.Sub))
+	h.Write([]byte("\000"))
+	h.Write([]byte(a.OutFmt))
+
+	key := fmt.Sprintf("%x", h.Sum(nil))
+	a.Key = key[:7]
+	return a.Key
 }
 
 func (a *Aggregator) AddOrCreate(key string, ts uint32, quantized uint, value float64) {
-	k := aggkey{
-		key,
-		quantized,
-	}
 	rangeTracker.Sample(ts)
-	proc, ok := a.aggregations[k]
+	aggByKey, ok := a.aggregations[quantized]
+	var proc Processor
 	if ok {
-		proc.Add(value, ts)
+		proc, ok = aggByKey[key]
+		if ok {
+			// if both levels exist, we can just add the value and that's it
+			proc.Add(value, ts)
+		}
 	} else {
+		// first level doesn't exist. create it and add the ts to the list
+		// (second level will be created below)
+		a.tsList = append(a.tsList, quantized)
+		if len(a.tsList) > 1 && a.tsList[len(a.tsList)-2] > quantized {
+			sort.Sort(TsSlice(a.tsList))
+		}
```

Comment:
Since we keep `a.tsList` sorted, it may be more efficient to use https://github.com/golang/go/wiki/SliceTricks#insert than to re-sort it, though I suppose in the vast majority of cases the newly-appended value would already be the greatest.

Reply:
re may be. yes, maybe..
re vast majority value already greatest, yes
doesn't seem worth digging into this deeper.